### PR TITLE
htcondor 1284 job epoch ads reading branch

### PR DESCRIPTION
**Pull request by ColeBollig**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_tools/history.cpp [hunk 1]
```diff
@@ -393,6 +398,28 @@ main(int argc, const char* argv[])
 			constraint.addCustomAND(where_expr.c_str());
 		}
 	}
+	else if (is_dash_arg_colon_prefix(argv[i], "scrape", &pcolon, 3)) { //TODO: Add flag to usage when ready to share with the world
+		//Designate reading job run instance ads from epoch directory
+		read_epoch_ads = true;
+		//Get epoch directory and validate passed arg is a directory
+		if ( (argc <= i+1)  || (*(argv[i+1]) == '-') ) { epochDirectory = param("JOB_EPOCH_INSTANCE_DIR"); }
+		else { epochDirectory = argv[i+1]; i++; }
```
Comment (johnkn) [inline src/condor_tools/history.cpp:406]: use ++i  here

File: src/condor_tools/history.cpp [hunk 2]
```diff
@@ -393,6 +398,28 @@ main(int argc, const char* argv[])
...
+		if ( epochDirectory != NULL ) {
+			StatInfo si(epochDirectory);
+			if (!si.IsDirectory() ) {
+				i--;
+				epochDirectory = param("JOB_EPOCH_INSTANCE_DIR");
```
Comment (johnkn) [inline src/condor_tools/history.cpp:411]: why param here? you just did that two lines above?

File: src/condor_tools/history.cpp [hunk 3]
```diff
@@ -1419,5 +1451,112 @@ static int set_print_mask_from_stream(
...
+	//If backwards then reverse the data structure
+	if (backwards) { std::reverse(epochFiles->begin(),epochFiles->end()); }
+}
+
+static void readHistoryFromEpochs(bool fileisuserlog, const char *JobHistoryFileName, const char* constraint, ExprTree *constraintExpr){
+	printHeader();
 
+	//If we were passed a specific file name then read that.
+	if (JobHistoryFileName) {
+		if (fileisuserlog) {
```
Comment (johnkn) [inline src/condor_tools/history.cpp:1494]: I don't think this is a option?

**Review by johnkn — CHANGES_REQUESTED** [review]

File: src/condor_tools/history.cpp (hunk 3 above)
Comment (ColeBollig) [reply src/condor_tools/history.cpp:1494]: Has been decided to be left in for now and dealt with in later changes regarding updating history file writing code

**Review by johnkn — APPROVED** [review]
Looks good for now.  we plan to moot some of this code very soon.

**Pull request merged**